src/geometry.F90

```diff
@@ -16,59 +17,129 @@ module geometry
-      if (actual_sense .neqv. specified_sense) then
-        in_cell = .false.
-        return
+    in_cell = .true.
+    do i = 1, size(c%rpn)
+      token = c%rpn(i)
+      if (token < OP_UNION) then
+        ! If the token is not an operator, evaluate the sense of particle with
+        ! respect to the surface and see if the token matches the sense. If the
+        ! particle's surface attribute is set and matches the token, that
+        ! overrides the determination based on sense().
+        if (token == p%surface) then
+          cycle
+        elseif (-token == p%surface) then
+          in_cell = .false.
+          exit
+        else
+          actual_sense = surfaces(abs(token))%obj%sense(&
+               p%coord(p%n_coord)%xyz, p%coord(p%n_coord)%uvw)
+          if (actual_sense .neqv. (token > 0)) then
+            in_cell = .false.
+            exit
+          end if
+        end if
       end if
-    end do SURFACE_LOOP
+    end do
+  end function simple_cell_contains
 
-    ! If we've reached here, then the sense matched on every surface or there
-    ! are no surfaces.
-    in_cell = .true.
+  pure function complex_cell_contains(c, p) result(in_cell)
+    type(Cell), intent(in) :: c
+    type(Particle), intent(in) :: p
+    logical :: in_cell
 
-  end function simple_cell_contains
+    integer :: i
+    integer :: token
+    logical :: b1, b2
+    integer :: i_stack
+    logical :: actual_sense    ! sense of particle wrt surface
+    logical :: stack(size(c%rpn))
+
+    i_stack = 0
+    do i = 1, size(c%rpn)
+      token = c%rpn(i)
+      if (token < OP_UNION) then
+        ! If the token is not an operator, evaluate the sense of particle with
+        ! respect to the surface and see if the token matches the sense. If the
+        ! particle's surface attribute is set and matches the token, that
+        ! overrides the determination based on sense().
+        i_stack = i_stack + 1
+        if (token == p%surface) then
+          stack(i_stack) = .true.
+        elseif (-token == p%surface) then
+          stack(i_stack) = .false.
+        else
+          actual_sense = surfaces(abs(token))%obj%sense(&
+               p%coord(p%n_coord)%xyz, p%coord(p%n_coord)%uvw)
+          stack(i_stack) = (actual_sense .eqv. (token > 0))
+        end if
+      else
+        ! If the token is a binary operator (intersection/union), apply it to
+        ! the last two items on the stack. If the token is a unary operator
+        ! (complement), apply it to the last item on the stack.
+        b1 = stack(i_stack)
+        select case (token)
+        case (OP_UNION)
+          b2 = stack(i_stack - 1)
+          stack(i_stack - 1) = b1 .or. b2
+          i_stack = i_stack - 1
+        case (OP_INTERSECTION)
+          b2 = stack(i_stack - 1)
+          stack(i_stack - 1) = b1 .and. b2
+          i_stack = i_stack - 1
+        case (OP_COMPLEMENT)
+          stack(i_stack) = .not. b1
+        end select
+      end if
+    end do
```

Comment:
You could reduce the depth of this logic a bit by removing the `if (token < OP_UNION)` and placing that content inside a `case default`